reco_utils/recommender/rbm/rbm.py

```diff
@@ -0,0 +1,946 @@
+                                      #percentage of the total training epoch when the step increases by 1
+
+        #Options to save the model for future use
+        self.save_model_ = save_model
+        self.save_path_ = save_path
+
+        self.debug_ = debug #if true, functions print their control paramters and/or outputs
+        self.with_metrics_ = with_metrics #compute msre and accuracy during training
+
+        #Initialize the start time
+        self.start_time = None
+
+        print("TensorFlow version: {}".format(tf.__version__))
+
+    #=========================
+    #Helper functions
+    #========================
+
+    # stateful time function
+    def time(self):
+        """
+        Time a particular section of the code - call this once to set the state somewhere
+        in the code, then call it again to return the elapsed time since last call.
+        Call again to set the time and so on...
+
+        Returns:
+             if timer started time in seconds since the last time time function was called
+        """
+
+        if self.start_time is None:
+            self.start_time = tm.time()
+            return False
+
+        else:
+            answer = tm.time() - self.start_time
+            # reset state
+            self.start_time = None
+            return answer
+
+        #else:
+        #    return None
+
+    #Binomial sampling
+    def B_sampling(self,pr):
+
+        '''
+        Binomial sampling of hidden units activations using a rejection method.
+
+        Args:
+            pr: input conditional probability
+            g:  uniform probability used for comparison
+
+        Returns:
+            h_sampled: sampled units. The value is 1 if pr>g and 0 otherwise.
+
+        Basic mechanics:
+            1) Extract a random number from a uniform distribution (g) and compare it with
+                the unit's probability (pr)
+
+            2) Choose 0 if pr<g, 1 otherwise. It is convenient to implement this condtion using
+               the relu function.
+
+        '''
+
+        np.random.seed(1)
+
+        #sample from a Bernoulli distribution with same dimensions as input distribution
+        g = tf.convert_to_tensor(np.random.uniform(size= pr.shape[1] ), dtype=tf.float32)
+
+        #sample the value of the hidden units
+        h_sampled = tf.nn.relu(tf.sign(pr-g) )
+
+        return h_sampled
+
+    #Multinomial sampling
+    def M_sampling(self, pr):
+
+        '''
+        Multinomial Sampling of ratings
+
+        Args:
+            pr:  a distributions of shape (m, n, r), where m is the number of examples, n the number
+                 of features and r the number of classes. pr needs to be normalized, i.e.
+                 sum_k p(k) = 1 for all m, at fixed n.
+            f:   normalized, uniform probability used for comparison.
+
+        Returns:
+            v_samp: an (m,n) array of sampled rankings from 1 to r .
+
+        Basic mechanics:
```

Comment:
here again, this should be before returns. Please take a look at google docstring format https://sphinxcontrib-napoleon.readthedocs.io/en/latest/example_google.html

Reply:
sure, I will reformat it in this way. Personally I find it more logic in this way: first define the variables than describe how it works, otherwise in the description you need to descfribe undefined objects.

Reply:
looking at the link you sent, I am not sure which option to compare to. For example, the format used in the example below seems similar to mine 

`def` example_generator(n):
    """Generators have a ``Yields`` section instead of a ``Returns`` section.

    Args:
        n (int): The upper limit of the range to generate, from 0 to `n` - 1.

    Yields:
        int: The next number in the range of 0 to `n` - 1.

    Examples:
        Examples should be written in doctest format, and should illustrate how
        to use the function.

        >>> print([i for i in example_generator(4)])
        [0, 1, 2, 3]'

Where Yields --> Returns and Examples --> Basic mechanics, that explains how the function is used.